src/Sparrow.Server/Allocators.cs

```diff
@@ -0,0 +1,681 @@
+    {
+        void Initialize<TAllocatorOptions>(TAllocatorOptions options) where TAllocatorOptions : struct, IAllocatorOptions;
+
+        TPointerType Allocate(int size);
+        void Release(ref TPointerType ptr);
+
+        void Renew();
+        void Reset();
+        void Dispose(bool disposing);
+
+        void LowMemory();
+        void LowMemoryOver();
+
+        bool IsThreadSafe { get; }
+    }
+
+    public interface IAllocator<T, TPointerType> : IAllocator
+        where T : struct, IAllocator
+        where TPointerType : struct, IPointerType
+    {
+        /// <summary>
+        /// This is the total amount of memory that has been allocated since the last Reset cycle.
+        /// </summary>
+        long TotalAllocated { get; }
+
+        /// <summary>
+        /// This is the total amount of memory currently either InUse or on hold on internal buffers since the last Reset Cycle.
+        /// </summary>
+        long Allocated { get; }
+
+        /// <summary>
+        /// This the is actual amount of memory that is on the customer hands at the current moment.
+        /// </summary>
+        long InUse { get; }
+
+        void Initialize(ref T allocator);
+
+        void Configure<TConfig>(ref T allocator, ref TConfig configuration) where TConfig : struct, IAllocatorOptions;
+
+        TPointerType Allocate(ref T allocator, int size);
+        void Release(ref T allocator, ref TPointerType ptr);
+        void Reset(ref T allocator);
+
+        void OnAllocate(ref T allocator, TPointerType ptr);
+        void OnRelease(ref T allocator, TPointerType ptr);
+
+        void Dispose(ref T allocator, bool disposing);
+    }
+
+    public static class Allocator
+    {
+        public static readonly SharedMultipleUseFlag LowMemoryFlag = new SharedMultipleUseFlag();
+    }
+
+    public sealed class Allocator<TAllocator> : IAllocatorComposer<Pointer>, IDisposable, ILowMemoryHandler
+        where TAllocator : struct, IAllocator<TAllocator, Pointer>
+    {
+        private TAllocator _allocator;
+        private readonly SingleUseFlag _disposeFlag = new SingleUseFlag();
+
+#if VALIDATE
+        private uint _generation;        
+#endif
+
+        ~Allocator()
+        {
+            if (_allocator is ILifecycleHandler<TAllocator> a)
+                a.BeforeFinalization(ref _allocator);
+
+            // We are not going to dispose twice, even if we hit here. 
+            if (_disposeFlag.Raise())
+                Dispose(false);
+        }
+
+        public void Initialize<TAllocatorOptions>(TAllocatorOptions options)
+            where TAllocatorOptions : struct, IAllocatorOptions
+        {
+
+            if (_allocator is ILifecycleHandler<TAllocator> a)
+                a.BeforeInitialize(ref _allocator);
+
+            _allocator.Initialize(ref _allocator);
+            _allocator.Configure(ref _allocator, ref options);
+
+#if VALIDATE
+            this._generation =  (uint)new Random().Next() & 0xFFF00000;
```

Comment:
Huh?

Reply:
Each `Reset` and `Renew` will increase the generation number. But in order to be able to track pointer crossovers between allocators, the first 24 bits are the allocator id and the rest are the actual generation. Both conditions are checked.